storage/integrate.go

```diff
@@ -0,0 +1,346 @@
+	if err != nil {
+		return 0, nil, nil, fmt.Errorf("failed to create range covering existing log: %w", err)
+	}
+
+	// Initialise a compact range representation, and verify the stored state.
+	r, err := baseRange.GetRootHash(nil)
+	if err != nil {
+		return 0, nil, nil, fmt.Errorf("invalid log state, unable to recalculate root: %w", err)
+	}
+	if len(entries) == 0 {
+		klog.V(1).Infof("Nothing to do.")
+		// Nothing to do, nothing done.
+		return fromSize, r, nil, nil
+	}
+
+	klog.V(1).Infof("Loaded state with roothash %x", r)
+	// Create a new compact range which represents the update to the tree
+	newRange := t.rf.NewEmptyRange(fromSize)
+	tc := newTileWriteCache(fromSize, t.getTile)
+	visitor := tc.Visitor(ctx)
+	for _, e := range entries {
+		lh := rfc6962.DefaultHasher.HashLeaf(e)
+		// Update range and set nodes
+		if err := newRange.Append(lh, visitor); err != nil {
+			return 0, nil, nil, fmt.Errorf("newRange.Append(): %v", err)
+		}
+
+	}
+	// Check whether the visitor had any problems building the update range
+	if err := tc.Err(); err != nil {
+		return 0, nil, nil, err
+	}
+
+	// Merge the update range into the old tree
+	if err := baseRange.AppendRange(newRange, visitor); err != nil {
+		return 0, nil, nil, fmt.Errorf("failed to merge new range onto existing log: %w", err)
+	}
+
+	// Check whether the visitor had any problems when merging the new range into the tree
+	if err := tc.Err(); err != nil {
+		return 0, nil, nil, err
+	}
+
+	// Calculate the new root hash - don't pass in the tileCache visitor here since
+	// this will construct any ephemeral nodes and we do not want to store those.
+	newRoot, err := baseRange.GetRootHash(nil)
+	if err != nil {
+		return 0, nil, nil, fmt.Errorf("failed to calculate new root hash: %w", err)
+	}
+
+	// All calculation is now complete, all that remains is to store the new
+	// tiles and updated log state.
+	klog.V(1).Infof("New log state: size 0x%x hash: %x", baseRange.End(), newRoot)
+
+	return baseRange.End(), newRoot, tc.Tiles(), nil
+
+}
+
+// tileReadCache is a structure which provides a very simple thread-safe read-through cache based on a map of tiles.
+type tileReadCache struct {
+	sync.RWMutex
+
+	entries  map[string]*populatedTile
+	getTiles func(ctx context.Context, tileIDs []TileID, treeSize uint64) ([]*api.HashTile, error)
+}
+
+func newTileReadCache(getTiles func(ctx context.Context, tileIDs []TileID, treeSize uint64) ([]*api.HashTile, error)) tileReadCache {
+	return tileReadCache{
+		entries:  make(map[string]*populatedTile),
+		getTiles: getTiles,
+	}
+}
+
+// Get returns a previously set tile and true, or, if no such tile is in the cache, attempt to fetch it.
+func (r *tileReadCache) Get(ctx context.Context, tileID TileID, treeSize uint64) (*populatedTile, error) {
+	r.Lock()
+	defer r.Unlock()
+	k := layout.TilePath(uint64(tileID.Level), tileID.Index, treeSize)
+	e, ok := r.entries[k]
+	if !ok {
+		t, err := r.getTiles(ctx, []TileID{tileID}, treeSize)
```

Comment:
Could be worth logging something here because there shouldn't be any misses if prewarm was used, and anyone not doing prewarm is missing out on performance wins.

Reply:
I'll pop a `V` log in there, I think there are legit times when this could happen during Integrate (e.g. when entirely new tiles are being created), but it'll be useful to peer in there and see what's happening.